sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1910,486 @@ def transform(self, X):
         """
         return _transform_selected(X, self._transform,
                                    self.categorical_features, copy=True)
+
+
+class QuantileTransformer(BaseEstimator, TransformerMixin):
+    """Transform features using quantiles information.
+
+    This method scales the features to follow a uniform or a normal
+    distribution. Therefore, for a given feature, this transformation tends
+    to spread out the most frequent values. It also reduces the impact of
+    (marginal) outliers: this is therefore a robust preprocessing scheme.
+
+    The transformation is applied on each feature independently.
+    The cumulative density function of a feature is used to project the
+    original values. Features values of new/unseen data that fall below
+    or above the fitted range will be mapped to the bounds of the output
+    distribution. Note that this transform is non-linear. It may distort linear
+    correlations between variables measured at the same scale but renders
+    variables measured at different scales more directly comparable.
+
+     Read more in the :ref:`User Guide <preprocessing_transformer>`.
+
+    Parameters
+    ----------
+    n_quantiles : int, optional (default=1000)
+        Number of quantiles to be computed. It corresponds to the number
+        of landmarks used to discretize the cumulative density function.
+
+    output_distribution : str, optional (default='uniform')
+        Marginal distribution for the transformed data. The choices are
+        'uniform' (default) or 'norm'.
+
+    ignore_implicit_zeros : bool, optional (default=False)
+        Apply only for sparse matrices. If True, the sparse entries of the
+        matrix are discarded to compute the quantile statistics. If false,
+        these entries are accounting for zeros.
+
+    subsample : int, optional (default=1e5)
+        Maximum number of samples used to estimate the quantiles.
+
+    smoothing_noise : None or float, optional (default=None)
```

Comment:
> What are your thoughts regarding this case which could be a use case, I think.

I find that case too contrived. If you can find real data (or invent something realistic) on which `QuantileTransformer` would be useful, but only especially effective with `smoothing_noise`, then I will admit it. Otherwise, showing that a single value lands up in a different place that matches our intuitions better does not show enough about the distribution of values and how that affects learning. If we have a real dataset we might find that we want a slightly different implementation, which is why I'm apprehensive about including it with only motivation from theory and toy examples. Did @ogrisel have something particular in mind?